README.md
💎⏫ Versionify
=======

Utility to simplify versioning in the app. Notifies current update status and provides download functionality via URL.

[![JitPack](https://jitpack.io/v/jdsdhp/versionify.svg)](https://jitpack.io/#jdsdhp/versionify) 
[![API](https://img.shields.io/badge/API-18%2B-red.svg?style=flat)](https://android-arsenal.com/api?level=18) 
[![License](https://img.shields.io/badge/License-MIT-blue.svg)](https://github.com/jdsdhp/versionify/blob/master/LICENSE) 
[![Twitter](https://img.shields.io/badge/Twitter-@jdsdhp-9C27B0.svg)](https://twitter.com/jdsdhp)

## Including in your project

#### Gradle

```gradle
allprojects  {
    repositories {
        ...
        maven { url 'https://jitpack.io' }
    }
}
dependencies {
    implementation 'com.github.jdsdhp:versionify:$version'
}
```

## Usage

### Maniefiest
Add the following permissions to the manifest file in your app.
```xml
<uses-permission android:name="android.permission.INTERNET" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```

### Kotlin
- Declare version property in your activity.
```kotlin
private var version: Version? = null
```
- Receive and handle you Version object from web service. For simplify this sample app, version object it's already created.
- Check the webservice version with current running app version and show dialog if is needed.
```kotlin
// Version sample from the web service.
version = Version(
    "1.2.3",
    "https://picsum.photos/200/300", //An image just for sample.
    """Include new features like:
        |- Add profile photo.
        |- New fancy fonts.
        """.trimMargin()
)
        
checkVersion(
    activity = this,
    version = version!!,
    title = "New Version",
    positiveText = "Download",
    negativeText = "Cancel",
    icon = R.drawable.ic_system_update,
    showAtMinimum = true   //By default is false.
)
```
- Handle needed permission to write into storage.
```kotlin
override fun onRequestPermissionsResult(
    requestCode: Int, permissions: Array<String>, grantResults: IntArray
) {
    var allGrant = true
    for (result in grantResults) if (result != PackageManager.PERMISSION_GRANTED) {
        allGrant = false
        break
    }
    if (allGrant) when (requestCode) {
        VERSION_REQUEST_CODE_W_EXT_STORAGE -> {
            version?.let { ver ->
                downloadVersion(
                    context = this,
                    downloadURL = ver.url,
                    title = getString(R.string.app_name),
                    description = getString(com.jesusd0897.versionify.R.string.version_downloading)
                        + " " + version!!.versionName,
                    downloadWay = DownloadWay.DOWNLOAD_MANAGER
                )
            }
                return
        }
    } else {
        Toast.makeText(this, "Permissions     Required!!", Toast.LENGTH_LONG).show()
        finish()
    }
}
```

## Sample project

It's very important to check out the sample app. Most techniques that you would want to implement are already implemented in the examples.

View the sample app's source code [here](https://github.com/jdsdhp/versionify/tree/master/app)

## TODO
* Add on Download complete Broadcast Receiver.

License
=======

    Copyright (c) 2020 jesusd0897.
    
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
Change Log
==========
Version 0.1.3 *(2020-08-12)*
----------------------------
Download ways added (Download Manger or Navigator Intent).

Version 0.1.2 *(2020-08-04)*
----------------------------
Change visibility of util classes to internal.

Version 0.1.0 *(2020-07-22)*
----------------------------
Initial release.

versionify/src/main/java/com/jesusd0897/versionify/Util.kt
/*
 * Copyright (c) 2020 jesusd0897.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jesusd0897.versionify

import android.app.Activity
import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import androidx.core.app.ActivityCompat
import com.google.gson.GsonBuilder
import com.jesusd0897.preferenza.consultPreference
import com.jesusd0897.preferenza.storePreference

private const val GENERAL_KEY_VERSION = "general_version"
private const val VERSION_KEY = "version_key"

enum class DownloadWay { DOWNLOAD_MANAGER, NAVIGATOR_INTENT }

/**
 * Request permission through Dialog.
 * @param activity           Activity who request.
 * @param manifestPermission Permission from manifest.
 * @param idPermission       Permission ID.
 */
internal fun requestDialogPermissions(
    activity: Activity,
    manifestPermission: String,
    idPermission: Int
) =
    ActivityCompat.requestPermissions(activity, arrayOf(manifestPermission), idPermission)

/**
 * Request permission through Dialog.
 * @param activity            Activity who request.
 * @param manifestPermissions Permissions from manifest.
 * @param idPermission        Permission ID.
 */
internal fun requestDialogPermissions(
    activity: Activity,
    manifestPermissions: Array<String>,
    idPermission: Int
) =
    ActivityCompat.requestPermissions(activity, manifestPermissions, idPermission)

/**
 * Provide a Version object wish contain full info of current supported
 * version by the web service.
 * @param context App context.
 */
fun apkVersion(context: Context): Version? {
    val versionString = consultPreference(context, GENERAL_KEY_VERSION, VERSION_KEY, "")
    if (versionString.isEmpty()) return null
    return try {
        val gson = GsonBuilder().serializeNulls().setPrettyPrinting().create()
        gson.fromJson(versionString, Version::class.java)
    } catch (ex: Exception) {
        ex.printStackTrace()
        null
    }
}

/**
 * Store a Version object wish contain full info of current supported
 * version by the web service.
 * @param context App context.
 * @param version App version from webservice to store into preferences.
 */
internal fun storeApkVersion(context: Context, version: Version) {
    val gson = GsonBuilder().serializeNulls().setPrettyPrinting().create()
    storePreference(context, GENERAL_KEY_VERSION, hashMapOf(VERSION_KEY to gson.toJson(version)))
}

/**
 * Provide stored version status.
 * @param context App context.
 * @param buildVersion App version name.
 */
fun isVersionOutdated(context: Context, buildVersion: String) =
    versionStatus(context, buildVersion) != VersionStatus.UPDATED

/**
 * Provide stored version status.
 * @param context App context.
 * @param buildVersion App version name.
 */
fun versionStatus(context: Context, buildVersion: String): VersionStatus {
    val fullVersion = apkVersion(context)
    if (fullVersion == null) return VersionStatus.BIG_CHANGE
    else {
        val cloudApkVersion = fullVersion.versionName.trim()
        if (buildVersion >= cloudApkVersion) return VersionStatus.UPDATED
        val installedDigits = buildVersion.split(".")
        val cloudDigits = cloudApkVersion.split(".")
        return when {
            installedDigits[0] < cloudDigits[0] -> VersionStatus.BIG_CHANGE
            installedDigits[1] < cloudDigits[1] -> VersionStatus.MEDIUM_CHANGE
            else -> VersionStatus.MINIMUM_CHANGE
        }
    }
}

fun downloadVersion(
    context: Context,
    downloadURL: String,
    title: String,
    description: String?,
    downloadWay: DownloadWay = DownloadWay.DOWNLOAD_MANAGER
) {
    when (downloadWay) {
        DownloadWay.DOWNLOAD_MANAGER -> downloadFileThroughDownloadManager(
            context, downloadURL, title, description
        )
        DownloadWay.NAVIGATOR_INTENT -> downloadFileThroughNavigatorIntent(context, downloadURL)
    }
}

/**
 * Ask the Download Manager to download a file from an URL.
 * @param context App context.
 * @param downloadURL Url to download apk file.
 * @param title Download manager notification title.
 * @param description Download manager notification description.
 */
private fun downloadFileThroughDownloadManager(
    context: Context,
    downloadURL: String,
    title: String,
    description: String?
) {
    val request = DownloadManager.Request(Uri.parse(downloadURL))
        .setAllowedNetworkTypes(DownloadManager.Request.NETWORK_WIFI or DownloadManager.Request.NETWORK_MOBILE)
        .setTitle(title)
        .setDescription(description)
        .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
        .setDestinationInExternalPublicDir(
            Environment.DIRECTORY_DOWNLOADS,
            title + "_" + System.currentTimeMillis() + ".apk"
        )
    request.allowScanningByMediaScanner()
    val dm = context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
    dm.enqueue(request)
}

/**
 * Send Download Intent to Navigator.
 * @param context App context.
 * @param downloadURL Url to download apk file.
 */
private fun downloadFileThroughNavigatorIntent(context: Context, downloadURL: String) =
    context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(downloadURL)))

versionify/src/main/java/com/jesusd0897/versionify/Alert.kt
/*
 * Copyright (c) 2020 jesusd0897.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jesusd0897.versionify

import android.Manifest
import android.app.Activity
import androidx.annotation.DrawableRes
import com.google.android.material.dialog.MaterialAlertDialogBuilder

const val VERSION_REQUEST_CODE_W_EXT_STORAGE = 4444

/**
 * Create a dialog builder for version control.
 * @param activity App context activity.
 * @param title Dialog title.
 * @param msg Dialog message.
 * @param positiveText  Actionable button text. By default os "Download".
 * @param icon Dialog icon.
 */
private fun versionWarningDialogBuilder(
    activity: Activity,
    title: String,
    msg: String,
    positiveText: String,
    @DrawableRes icon: Int? = null
) = MaterialAlertDialogBuilder(activity)
    .setTitle(title)
    .setMessage(msg)
    .setPositiveButton(positiveText) { _, _ ->
        requestDialogPermissions(
            activity,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            VERSION_REQUEST_CODE_W_EXT_STORAGE
        )
    }.apply {
        icon?.let { this.setIcon(it) }
    }

private fun versionWarningDialogIfNeeded(
    activity: Activity,
    newVersion: Version,
    status: VersionStatus,
    title: String,
    positiveText: String,
    negativeText: String,
    @DrawableRes icon: Int? = null,
    showAtMinimum: Boolean = false
) {
    if (status != VersionStatus.UPDATED && (status != VersionStatus.MINIMUM_CHANGE || showAtMinimum)) {
        val message =
            activity.getString(status.message) +
                    ": v." + "${newVersion.versionName}\n•••••••\n" +
                    "${newVersion.description}"
        val builder = versionWarningDialogBuilder(
            activity, title, message, positiveText, icon
        )
        if (status == VersionStatus.BIG_CHANGE) builder.setCancelable(false)
        else builder.setNegativeButton(negativeText, null)
        builder.show()
    }
}

/**
 * @param activity Context activity.
 * @param version Web service version.
 * @param title Dialog title.
 * @param positiveText  Actionable button text. By default os "Download".
 * @param negativeText Cancelable button text. By default os "Cancel".
 * @param icon Dialog icon.
 * @param showAtMinimum Indicates if it should be shown when it is a minimum change. By default is false.
 */
fun checkVersion(
    activity: Activity,
    version: Version,
    title: String? = null,
    positiveText: String? = null,
    negativeText: String? = null,
    @DrawableRes icon: Int? = null,
    showAtMinimum: Boolean = false
) {
    storeApkVersion(activity, version)
    val buildVersionName =
        activity.packageManager.getPackageInfo(activity.packageName, 0).versionName
    val status = versionStatus(activity, buildVersionName)
    versionWarningDialogIfNeeded(
        activity,
        version,
        status,
        title ?: activity.getString(R.string.version_warning_title),
        positiveText ?: activity.getString(R.string.version_download),
        negativeText ?: activity.getString(R.string.version_cancel),
        icon,
        showAtMinimum
    )
}

versionify/src/main/java/com/jesusd0897/versionify/Version.kt
/*
 * Copyright (c) 2020 jesusd0897.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jesusd0897.versionify

import android.os.Parcelable
import androidx.annotation.StringRes
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import kotlinx.android.parcel.Parcelize

private const val APP_KEY_TAG = "app_key"
private const val APP_KEY = "citas"

private const val VERSION_TAG = "version"
private const val DOWNLOAD_TAG = "descarga"
private const val DESCRIPTION_TAG = "funcionalidades"

@Parcelize
data class AppKey(@Expose @SerializedName(APP_KEY_TAG) var appKey: String = APP_KEY) : Parcelable

@Parcelize
data class Version(
    @Expose @SerializedName(VERSION_TAG) var versionName: String,
    @Expose @SerializedName(DOWNLOAD_TAG) var url: String,
    @Expose @SerializedName(DESCRIPTION_TAG) var description: String?
) : Parcelable

enum class VersionStatus(@StringRes val message: Int) {
    UPDATED(R.string.version_updated_msg),
    MINIMUM_CHANGE(R.string.version_minimum_change_msg),
    MEDIUM_CHANGE(R.string.version_medium_change_msg),
    BIG_CHANGE(R.string.version_big_change_msg),
}

app/src/main/java/com/jesusd0897/versionify/sample/MainActivity.kt
/*
 * Copyright (c) 2020 jesusd0897.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.jesusd0897.versionify.sample

import android.content.pm.PackageManager
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.jesusd0897.versionify.*

class MainActivity : AppCompatActivity() {

    private var version: Version? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Version from the web service
        version = Version(
            "1.2.3",
            //"https://github.com/skydoves/Pokedex/releases/download/1.0.1/app-debug.apk",
            "https://picsum.photos/200/300",
            """Include new features like:
                |- Add profile photo.
                |- New fancy fonts.
            """.trimMargin()
        )

        checkVersion(
            activity = this,
            version = version!!,
            title = "New Version",
            positiveText = "Download",
            negativeText = "Cancel",
            icon = R.drawable.ic_system_update,
            showAtMinimum = true   //By default is false.
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        var allGrant = true
        for (result in grantResults) if (result != PackageManager.PERMISSION_GRANTED) {
            allGrant = false
            break
        }
        if (allGrant) when (requestCode) {
            VERSION_REQUEST_CODE_W_EXT_STORAGE -> {
                version?.let { ver ->
                    downloadVersion(
                        context = this,
                        downloadURL = ver.url,
                        title = getString(R.string.app_name),
                        description = getString(com.jesusd0897.versionify.R.string.version_downloading)
                                + " " + version!!.versionName,
                        downloadWay = DownloadWay.DOWNLOAD_MANAGER
                    )
                }
                return
            }
        } else {
            Toast.makeText(this, "Permissions Required!!", Toast.LENGTH_LONG).show()
            finish()
        }
    }

}